clang/lib/CodeGen/Targets/AArch64.cpp

```diff
@@ -812,6 +819,24 @@ Address AArch64ABIInfo::EmitMSVAArg(CodeGenFunction &CGF, Address VAListAddr,
                           /*allowHigherAlign*/ false);
 }
 
+void AArch64TargetCodeGenInfo::checkFunctionCallABI(
+    CodeGenModule &CGM, SourceLocation CallLoc, const FunctionDecl *Caller,
+    const FunctionDecl *Callee, const CallArgList &Args) const {
+  if (!Callee->hasAttr<AlwaysInlineAttr>())
+    return;
+
+  auto CalleeStreamingMode = Sema::getArmStreamingFnType(Callee);
+  auto CallerStreamingMode = Sema::getArmStreamingFnType(Caller);
+
+  // The caller can inline the callee if their streaming modes match or the
+  // callee is streaming compatible
+  if (CalleeStreamingMode != CallerStreamingMode &&
```

Comment:
It is valid to inline Callees with ArmStreaming or ArmStreamingCompatible into Callers with the __arm_locally_streaming keyword? What happens currently when you try this?

Reply:
It is indeed valid to inline a streaming function into a locally streaming function since at the inline point streaming is guaranteed to be on. We don't need the streaming mode save and restore that would come from a locally streaming function's prologue and epilogue.